### DISCUSSION TITLE
useState not working in .on().subscribe()

### DISCUSSION THREAD

**Discussion by HassanIyan** [supabase/supabase#3887] [Questions] [answered]
i am trying to create a realtime app using supabase and react. but i am facing a problem. i am using the code below

useState for post
`const [post, setPost] = useState([])`

realtime task code
```
const mySubscription = supabase
  .from('post')
  .on('INSERT', payload => {
    setPost(payload)
    console.log(post)
    console.log(payload)
  })
  .subscribe()
```

when i console log the payload i get the array as expected,
but if a console.log on post i get a empty array. is this a bug or am i doing it wrong??

**Comment by joshenlim** [comment, marked as answer]
Hey @HassanIyan, this is because the setting of state (`setPost`) in React is asynchronous in nature, hence why console logging `post` immediately after calling setPost returns stale data.

You can refer to a similar SO thread here for more details: 
https://stackoverflow.com/questions/54069253/usestate-set-method-not-reflecting-change-immediately

To better visualize the updating of your `post` state, you can try rendering in JSX the `post` variable instead like so `<div>{JSON.stringify(post)}</div>`, that way you'll be able to see the data update

Separately, it might also be good to know that the following code snippet will not work as well, for the same reason that the useState hook is asynchronous

```
const [ data, setData ] = useState('old')

const function1 = () => {
  setData('hello')
  function2()
}

const function2 = () => {
  console.log(data) // Will show 'old', and not 'hello'
}
```

If you do end up in a scenario where you need this ^ to work, then using `useRef` might be a possible workaround though not the best solution (equivalent of declaring a variable via `let` and mutating it)

```
const dataRef = useRef(null)
dataRef.current = 'old'

const function1 = () => {
  dataRef.current = 'hello')
  function2()
}

const function2 = () => {
  console.log(dataRef.current) // will show 'hello'
}
```
